### PR TITLE
Add flatten composite rule

### PR DESCRIPTION
**Pull request by xysheng-baidu**
### PR types

Others
### PR changes

Others
### Describe

add composite rule for flatten

### CODE REVIEW

File: python/paddle/incubate/autograd/composite_rules.py [hunk 1]
```diff
@@ -178,3 +178,40 @@ def mean_composite(x, axis, keepdim):
...
+    max = dim_post_expr - 1
+    assert not (dim < min or dim > max)
+    if dim < 0:
+        dim += dim_post_expr
+    return dim
+
+
+@REGISTER_COMPOSITE('flatten_contiguous_range')
+def flatten_contiguous_range_composite(x, start_axis, stop_axis):
+    """define composite rule of op flatten, flatten_contiguous_range -> flatten"""
```
Comment (cyber-pioneer) [inline python/paddle/incubate/autograd/composite_rules.py]: 记录拆分逻辑

Comment (xysheng-baidu) [reply python/paddle/incubate/autograd/composite_rules.py]: 已添加拆分逻辑。

File: python/paddle/incubate/autograd/composite_rules.py [hunk 2]
```diff
@@ -182,6 +182,48 @@ def mean_composite(x, axis, keepdim):
...
+@REGISTER_COMPOSITE('flatten_contiguous_range')
+def flatten_contiguous_range_composite(x, start_axis, stop_axis):
+    """
+    define composite rule of op flatten, flatten_contiguous_range -> flatten.
+    xshape is the dim with 0 added to the front of x, keep the shape information of x to calculate the grad.
+    shape_out is the parameter of reshape, get from start_axis and stop_axis.
+    out = reshape(x, shape=shape_out), xshape
+    """
+    shape_in = x.shape
+    shape_x_out: List[int] = [0]
```
Comment (cyber-pioneer) [inline python/paddle/incubate/autograd/composite_rules.py]: 这里语法确认一下

Comment (xysheng-baidu) [reply python/paddle/incubate/autograd/composite_rules.py]: @cyber-pioneer 改成通用写法了，测试通过。

File: python/paddle/fluid/tests/unittests/test_flatten_contiguous_range_op.py [hunk 3]
```diff
@@ -25,21 +25,25 @@ def setUp(self):
         self.python_api = paddle.flatten
         self.python_out_sig = ["Out"]
         self.op_type = "flatten_contiguous_range"
+        self.prim_op_type = "comp"
         self.start_axis = 0
         self.stop_axis = -1
         self.init_test_case()
         self.inputs = {"X": np.random.random(self.in_shape).astype("float64")}
         self.init_attrs()
+        self.enable_cinn = False
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_flatten_contiguous_range_op.py:34]: 子类的CINN测试都过不了吗

Comment (xysheng-baidu) [reply python/paddle/fluid/tests/unittests/test_flatten_contiguous_range_op.py:34]: 是的，没有能过的。

**Review by cyber-pioneer — DISMISSED** [review]

**Comment by cyber-pioneer** [comment]
LGTM

File: python/paddle/incubate/autograd/composite_rules.py [hunk 4]
```diff
@@ -191,6 +191,35 @@ def mean_composite(x, axis, keepdim):
...
+    slice_numel = 1
+    for i in range(start_dim, end_dim + 1):
+        slice_numel *= shape_in[i]
+    shape_out = []
+    for i in range(start_dim):
+        shape_out.append(shape_in[i])
+    shape_out.append(slice_numel)
+    for i in range(end_dim + 1, len(shape_in)):
+        shape_out.append(shape_in[i])
+    return reshape(x, shape=shape_out), xshape
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/composite_rules.py]: Can this op doesn't return xshape?

**Review by JiabinYang — COMMENTED** [review]
one comment

File: python/paddle/incubate/autograd/composite_rules.py (hunk 4 above)
Comment (xysheng-baidu) [reply python/paddle/incubate/autograd/composite_rules.py]: There is xshape in the yaml definition, and need to use xshape when calculate grad.

Comment (xysheng-baidu) [reply python/paddle/incubate/autograd/composite_rules.py]: The yaml defineition :
https://github.com/PaddlePaddle/Paddle/blob/6786c012c144c29f976fc00e0e62f55dbb2d34ec/paddle/phi/api/yaml/legacy_ops.yaml#L644

**Review by JiabinYang — COMMENTED** [review]
some coments

File: python/paddle/incubate/autograd/composite_rules.py (hunk 4 above)
Comment (JiabinYang) [reply python/paddle/incubate/autograd/composite_rules.py]: Yes, but for this case, which is a little special, CINN can't optimize xshape and actually it doesn't need xshape for backward pass. So, it's better to remove xshape usage if it's possible

**Review by JiabinYang — COMMENTED** [review]
one comment

**Review by JiabinYang — DISMISSED** [review]
LGTM

**Comment by xysheng-baidu** [comment]
@JiabinYang remove xshape and return none, passed the test.

**Review by JiabinYang — APPROVED** [review]
LGTM

**Pull request merged**